modelforge/potential/aimnet2.py

```diff
@@ -168,54 +169,445 @@ def compute_properties(
+        ----------
+        per_atom_feature_tensor : torch.Tensor
+            The feature tensor (either atomic embeddings or repeated partial charges).
+        pairlist : torch.Tensor
+            The list of atom pairs.
+        f_ij_cutoff : torch.Tensor
+            The cutoff function applied to the radial symmetry functions.
+        r_ij : torch.Tensor
+            The displacement vectors between atom pairs.
+        linear_transform : nn.Module
+            The linear transformation to apply to the features.
+
+        Returns
+        -------
+        Tuple[torch.Tensor, torch.Tensor]
+            Radial and vector contributions.
+        """
+
+        idx_j = pairlist[1]
+
+        # Calculate the unit vector u_ij
+        r_ij_norm = torch.norm(r_ij, dim=1, keepdim=True)  # Shape: (num_atom_pairs, 1)
+        u_ij = r_ij / r_ij_norm  # Shape: (num_atom_pairs, 3)
+
+        # Step 1: Radial Contributions Calculation (Equation 4)
+        proto_v_r_a = (
+            f_ij_cutoff * per_atom_feature_tensor[idx_j]
+        )  # Shape: (num_atom_pairs, nr_of_features)
+
+        # Initialize tensor to accumulate radial contributions for each atom
+        radial_contributions = torch.zeros(
+            (per_atom_feature_tensor.shape[0], self.number_of_per_atom_features),
+            device=per_atom_feature_tensor.device,
+            dtype=per_atom_feature_tensor.dtype,
+        )  # Shape: (num_of_atoms, nr_of_features)
+
+        # Accumulate the radial contributions using index_add_
+        radial_contributions.index_add_(0, idx_j, proto_v_r_a)
+
+        # Step 2: Vector Contributions Calculation (Equation 5)
+        # First, calculate the directional component by multiplying g_ij with u_ij
+        vector_prot_step1 = u_ij.unsqueeze(-1) * f_ij_cutoff.unsqueeze(
+            -2
+        )  # Shape: (num_atom_pairs, 3, nr_of_features)
+
+        # Next, multiply this result by the input of atom j
+        vector_prot_step2 = vector_prot_step1 * per_atom_feature_tensor[
+            idx_j
+        ].unsqueeze(
+            1
+        )  # Shape: (num_atom_pairs, 3, nr_of_features)
+
+        # Sum over the last dimension (nr_of_features) to reduce it
+        vector_prot_step2 = vector_prot_step2.sum(dim=-1)  # Shape: (num_atom_pairs, 3)
+
+        # Initialize tensor to accumulate vector contributions for each atom
+        vector_contributions = torch.zeros(
+            per_atom_feature_tensor.shape[0], 3
+        )  # Shape: (num_of_atoms, 3)
+
+        # Accumulate the vector contributions using index_add_
+        vector_contributions.index_add_(0, idx_j, vector_prot_step2)
+
+        # Step 3: Compute the Euclidean Norm for each atom
+        vector_norms = torch.norm(
+            vector_contributions, p=2, dim=1
+        )  # Shape: (num_of_atoms,)
+
+        return radial_contributions, vector_norms
+
+
+class FirstMessageModule(BaseMessageModule):
+    def __init__(self, number_of_per_atom_features: int):
```

Comment:
This constructor calls the base class constructor which already calls:

```
self.linear_transform_embeddings = nn.Linear(
    number_of_per_atom_features, number_of_per_atom_features
)
```
You should be able to remove this.

However, seeing that the SubsequentMessageModule also does this, it looks like we're allocating resources to the `linear_transform_charges` tensor here when we shouldn't be.